src/ColorVectorSpace.jl

```diff
@@ -207,6 +207,35 @@ for op in unaryOps
     @eval ($op)(c::AbstractGray) = $op(gray(c))
 end
 
+function var(A::AbstractArray{C}; kwargs...) where C<:AbstractGray
+    imgc = channelview(A)
```

Comment:
Oops, `channelview` comes from `ImageCore`, which is so unfortunate because we might not want to introduce `ImageCore` as a dependency on `ColorVectorSpace`; instead, we prefer `using ColorVectorSpace` in `ImageCore`. https://github.com/JuliaGraphics/ColorTypes.jl/pull/160#issuecomment-583868311

But given that the best place I have in mind for `var` is `ColorVectorSpace`, adding `ImageCore` to `ColorVectorSpace` seems a must choice.

cc: @timholy

Reply:
I don't think we need the `channelview` here just to implement `var`.
However, if there are many such methods using `channelview`, I think that the versions using `channelview` should be extended in `ImageCore.jl`.

BTW, I do not fully understand why `var` of vectors should be element-wise (channel-wise).:sweat_smile:

Reply:
> I don't think we need the channelview here just to implement var.

No, we don't necessarily need, but it would make trivial things less trivial. Another choice is to move these functions to `ImageDistances`, rename it to `ImageStats` and absorb `ImageQualityIndexes` as well.

---

> I do not fully understand why var of vectors should be element-wise (channel-wise)

Generally, for non-gray image `x` we don't know how to combine those channels. But yes, RGB image can be one special case that most people just treat it as `3*m*n` array.

For example, PSNR treats RGB and other Color3 images differently. https://github.com/JuliaImages/ImageQualityIndexes.jl/blob/master/src/psnr.jl#L13-L21

From [wikipedia](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio)

> Alternately, for color images the image is converted to a different color space and PSNR is reported against each channel of that color space, e.g, YCbCr or HSL


I agree that we may need one more method for `AbstractRGB` images.

Reply:
*IMO*, if there are no clear rules about combining channels, ColorVectorSpace should only support a single-channel (grayscale) arrays.

Reply:
Basically I agree with you. From the [commit history](https://github.com/JuliaImages/Images.jl/commit/664cb157f48110e4622591c6389b1139a3c19dc5) I think deprecating it would be acceptable.

But at least for now the best choice is to move it here and to add a deprecation warning.

My suggestion is to keep `AbstractGray` and `AbstractRGB` and to throw depwarn for other `Colorant` types unless with have a specific usage for them.

Reply:
You may think my comments are inconsistent, but I'm not negative in channel-wise variance itself. I just care about the problem of a lower-level package depending on higher-level packages.
By analogy with the multivariate statistics, I think the current behavior is reasonable. So I don't think we need the deprecation.
```julia
julia> using Statistics

julia> wbb = [[1.0,1.0,1.0], [0.0,0.0,0.0], [0.0,0.0,0.0]]
3-element Array{Array{Float64,1},1}:
 [1.0, 1.0, 1.0]
 [0.0, 0.0, 0.0]
 [0.0, 0.0, 0.0]

julia> rgb = [[1.0,0.0,0.0], [0.0,1.0,0.0], [0.0,0.0,1.0]]
3-element Array{Array{Float64,1},1}:
 [1.0, 0.0, 0.0]
 [0.0, 1.0, 0.0]
 [0.0, 0.0, 1.0]

julia> var(wbb)
3-element Array{Float64,1}:
 0.33333333333333337
 0.33333333333333337
 0.33333333333333337

julia> var(rgb)
3-element Array{Float64,1}:
 0.33333333333333337
 0.33333333333333337
 0.33333333333333337
```

Reply:
> I think the current behavior is reasonable

It makes sense. I had a feeling about that but didn't check that previously.

---

Corrects me if I don't understand it right, let me try to state the roles of ImageCore, Colors and ColorVectorSpaces

* `Colors` provides fundamental color operations on pixel-level as a delicately designed art.
* `ColorVectorSpaces` provides extensive glue support on pixel-level for methods from other packages, mostly `Base`.
* `ImageCore` is like `Colors` but it provides operations on array-level.

Revisiting ColorVectorSpaces, I don't see any method on Array types.

I now think I was wrong saying ColorVectorSpace is the most suitable place for `var` and `std` in https://github.com/JuliaImages/Images.jl/issues/872 we miss a ColorVectorSpace-like glue package on array level for `Base`, `Statistics` and etc.

For the array-level glue package, we only have `ImageDistances` and `ImageAxes`

---

For this PR, one thing for sure is that `complement` still needs to stay here.

Reply:
I agree for the most part of the roles.
However, I think the role of ColorVectorSpace, as the name implies, is to define a vector space for colors, i.e. to introduce linearity. The introduction of linearity allows for the addition, subtraction and uniform scaling of colors. So we can define the "mean" of the colors. In fact, we haven't defined the `mean` method in ColorVectorSpace, but the `Statistics` module provides it.

Similarly, `var` and` std` can be defined systematically. However, there are two problems:
1. `Statistics` module does not fully support colors.
2. The definition of the "difference" from the mean color is ambiguous.

My comment above
> BTW, I do not fully understand why var of vectors should be element-wise (channel-wise).

means "2.". I don't think "pixel-level" or "array-level" is the essence of the problem. Therefore, I think that defining "default" `var` in ColorVectorSpace itself is fine, but I disagree with using `channelview` just for that.

---

> For this PR, one thing for sure is that complement can stay in this package.
>
> But...I now think it might be better to stay in Colors...

The current `complement` implicitly assumes linearity. For example, what is the "complement" of `Lab(20, 30, 40)`? The complementary color is a more general (i.e. ambiguous) concept.

Reply:
> I think the role of ColorVectorSpace, as the name implies, is to define a vector space for colors, i.e. to introduce linearity.

Yes, I understand this is the motivation of this package. Still, I tried to avoid saying this when I described the roles because I thought it doesn't help to solve the current issue. 

For example, by saying this, any function in ImageDistances has their right to stay here, but introducing any of them would trigger such a discussion thread.

> Therefore, I think that defining "default" (fallback) var in ColorVectorSpace itself is fine, but I disagree with using channelview just for that.

The pixel/array level difference is the only thing that I can speak of the low/high level abstraction. Since both packages don't strictly rely on each other, we can't say that one is more fundamental than the other.

As I said, not using channelview would make trivial things less trivial, it's not very replicable. If we can't use `channelview`, then _IMO_ it basically suggests that either `ColorVectorSpace` isn't a good place for `var` or `channelview` should be moved to `Colors.jl`. However, the status quo is that `channelview` should still stay in `ImageCore`.

Thus, I prefer to create an ImageBase package to fit these kinds of needs for array-level operations. Indeed, if there's no objection about that, I can do it myself.

Reply:
> For example, by saying this, any function in ImageDistances has their right to stay here, but introducing any of them would trigger such a discussion thread.

The difference from Statistics is that Distance is not a stdlib. I do not object to creating a new package to reduce the size of the package. (I think ImageCore and ImageBase can be confusing, though.)

> The pixel/array level difference is the only thing that I can speak of the low/high level abstraction.

ColorVectorSpace defines the color arithmetic. Perhaps the "pixel-level" you said might mean that, but the arithmetic should be independent of the shape of arrays. In short, the problem with `var` is a matter of what the type/value of "square error of colors" should be.

> As I said, not using channelview would make trivial things less trivial, it's not very replicable.

What you said is correct, but no channel splitter is needed to calculate `var`. Essentially, we need the following methods:
```julia
mapreduce(color->mapc(abs2, color - mean_color), +, A)
```
Handling the keyword arguments (especially `dims`) is annoying, but I think it's better than depending on the inside of Statistics.jl. Although I have no strong opinion on the location of `var`, we may obtain the "speed" as a practical effect by not using `channelview`. :smiling_imp: